crates/accelerate/src/quantum_info/versor_u2.rs

```diff
@@ -0,0 +1,554 @@
+    };
+}
+
+impl ::std::ops::Mul for VersorSU2 {
+    type Output = VersorSU2;
+    fn mul(self, other: VersorSU2) -> Self::Output {
+        VersorSU2(self.0 * other.0)
+    }
+}
+impl_mul_refs!(VersorSU2, VersorSU2);
+
+impl ::std::ops::Mul for VersorU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorU2) -> Self::Output {
+        VersorU2 {
+            phase: self.phase + other.phase,
+            su2: self.su2 * other.su2,
+        }
+    }
+}
+impl_mul_refs!(VersorU2, VersorU2);
+
+impl ::std::ops::Mul<VersorU2> for VersorSU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorU2) -> VersorU2 {
+        VersorU2 {
+            phase: other.phase,
+            su2: self * other.su2,
+        }
+    }
+}
+impl_mul_refs!(VersorU2, VersorSU2);
+
+impl ::std::ops::Mul<VersorSU2> for VersorU2 {
+    type Output = VersorU2;
+    fn mul(self, other: VersorSU2) -> VersorU2 {
+        VersorU2 {
+            phase: self.phase,
+            su2: self.su2 * other,
+        }
+    }
+}
+impl_mul_refs!(VersorSU2, VersorU2);
+
+/// A module-internal trait to simplify the code-generation of both the dynamic `ndarray` and the
+/// static `&[[Complex64; 2]; 2]` and `Matrix2` paths.  Rather than making the user care
+/// about importing it, we just expose the concretised methods using it through `VersorU2`.
+trait Matrix1q {
+    fn get(&self, row: usize, col: usize) -> Complex64;
+}
+impl<const N: usize, const M: usize> Matrix1q for [[Complex64; N]; M] {
+    #[inline(always)]
+    fn get(&self, row: usize, col: usize) -> Complex64 {
+        self[row][col]
+    }
+}
+impl Matrix1q for ArrayView2<'_, Complex64> {
+    #[inline(always)]
+    fn get(&self, row: usize, col: usize) -> Complex64 {
+        self[(row, col)]
+    }
+}
+impl Matrix1q for Matrix2<Complex64> {
+    #[inline(always)]
+    fn get(&self, row: usize, col: usize) -> Complex64 {
+        self[(row, col)]
+    }
+}
+
+/// Calculate the squared Frobenius distance `|| A+ A - 1 ||_2^2`.
+///
+/// As a rule of thumb, if `A+ A = 1 + eps`, then the Frobenius distance from the identity will be
+/// `sqrt(2)*eps`, and this function would return `2 * eps*eps`, up to round-off error.
+fn unitary_frobenius_distance_square<M: Matrix1q>(matrix: &M) -> f64 {
+    let topleft = matrix.get(0, 0).norm_sqr() + matrix.get(0, 1).norm_sqr();
+    let botright = matrix.get(1, 0).norm_sqr() + matrix.get(1, 1).norm_sqr();
+    let off =
+        matrix.get(0, 0) * matrix.get(1, 0).conj() + matrix.get(0, 1) * matrix.get(1, 1).conj();
+    (topleft - 1.).powi(2) + (botright - 1.).powi(2) + 2. * off.norm_sqr()
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use approx::AbsDiffEq;
+    use ndarray::aview2;
+    use qiskit_circuit::operations::{Operation, Param, StandardGate, STANDARD_GATE_SIZE};
+
+    fn all_1q_gates() -> Vec<StandardGate> {
+        (0..STANDARD_GATE_SIZE as u8)
+            .filter_map(|x| {
+                ::bytemuck::checked::try_cast::<_, StandardGate>(x)
```

Comment:
I actually have code somewhere that I added a `impl TryFrom<u8> for StandardGate` and just made a static  `[StandardGate; STANDARD_GATE_SIZE]` array to do the conversion. It was part of some code I didn't end up finish or using for anything. But I can dig up the branch and push that as a standalone if you think it would be useful (I added it for a similar use case to this).

Reply:
`TryFrom<u8> for StandardGate` _feels_ wrong to me, but not super strongly.  It feels nicer somehow using `bytemuck` to say "I know I'm doing something a little bit odd", since it feels like we're kind of conjuring semantics out of nowhere - the fact that `StandardGate` is backed by integers is (imo) an implementation detail, albeit a fixed one.

If we wanted to expose an iterator over all gates, I'd just expose your `[StandardGate; Standard_GATE_SIZE]` as a `static` publicly, then you can iterate through it directly.  That's what I did with `BitTerm` in the tests of `SparseObservable`'s lookup tables.